### ISSUE TITLE
[BUG] Models are slow to train

### ISSUE THREAD

**Issue by benfred** [NVIDIA-Merlin/models#208]
### Bug description

Models using negative sampling are very slow to train. As an example, using ` merlin.models.tf.MatrixFactorizationModel` takes about 14 hours an epoch to train on the movielens dataset. For comparison,  each epoch takes implicit about 250ms on the same dataset - meaning implicit is over 190,000x faster.

### Steps/Code to reproduce bug

With the code from https://github.com/NVIDIA-Merlin/models/pull/207 :

```python
# get the movielens dataset
from merlin.models.utils.data_etl_utils import get_movielens
train, valid = get_movielens(variant="ml-25m")

# create a simple TF MatrixFactorizationModel on the movielens dataset
import merlin.models.tf as ml
model = ml.MatrixFactorizationModel(train.schema, dim=64)
model.compile(optimizer="adam", run_eagerly=True)

# create a dataloader: note this should be unnecessary, and we should be able
# to fit the model directly from the dataset - but it doesn't seem to work
schema = train.schema
import merlin.models.tf.dataset as tf_dataloader
def get_dataloader(paths_or_dataset, batch_size=4096, shuffle=True):
    dataloader = tf_dataloader.Dataset(
        paths_or_dataset,
        batch_size=batch_size,
        label_names=["rating_binary"],
        shuffle=shuffle,
        schema=schema,
    )
    return dataloader.map(lambda X, y: (X, tf.reshape(y, (-1,))))
train_loader = get_dataloader(train, shuffle=True)

# train the model: this will take several days to finish
model.fit(train_loader, batch_size=1024, epochs=15)
```

### Expected behavior

I would expect merlin-models to be able to train a simple model on the movielens dataset in less than a minute per epoch.

**Comment by benfred** [comment]
fwiw, 99% of the time seems to be in calcualting the eval_metrics: 

![slow_model](https://user-images.githubusercontent.com/69536/156112191-af42b6c8-8fff-4cd9-bc70-5c4f7867326a.svg)

**Comment by marcromeyn** [comment]
Thanks for filing this! We are currently working on this, we want to allow to calculate computationally intensive metrics only every _n_ steps, instead of for each step.

**Comment by benfred** [comment]
That will only speed this up `N` times though right?  I'm thinking that even if we speed this up 100x , this will still be pretty slow.

Can we also maybe make some other changes?
* Limit the number of expensive metrics we're calculating. Do we need all of NDCG@K / MAP@K / P@K  / recall@k ? (also I think we're doing this for multiple values of k ?)
* What about just defaulting to cheaper metrics to calculate like AUC. With pair-wise losses this can be calculated basically for free (its the percentage of times the positive sample has a higher score than the negative sample_

**Comment by sararb** [comment]
Thank you for the catch! We are working on different solutions to optimize calculating metrics for retrieval models: 
- In this PR #189 : we decoupled the top-k extraction and calculating metrics so as we do the top-k extraction just once (with the maximum k from different thresholds defined in metrics).  We also exposed a boolean flag `compute_train_metrics` to deactivate computing metrics during training and only gather validation metrics after each epoch.

- We are also investigating other sources of optimization: #210 and #202. 

- Setting a default cheap metric for training could also be an option to add.

**Comment by rnyak** [comment]
@marcromeyn @sararb  I also shared in the slack channel that tf_dataloader is slower than KerasSequenceLoader. Ben  can you test this instead:

```
def get_dataloader(paths_or_dataset, batch_size=4096, shuffle=True):
    dataloader = KerasSequenceLoader(
        paths_or_dataset,
        batch_size=batch_size,
        label_names=["rating_binary"],
        shuffle=shuffle,
        schema=schema,
    )
    return dataloader.map(lambda X, y: (X, tf.reshape(y, (-1,))))
```

**Comment by rnyak** [comment]
> That will only speed this up `N` times though right? I'm thinking that even if we speed this up 100x , this will still be pretty slow.
> 
> Can we also maybe make some other changes?
> 
> * Limit the number of expensive metrics we're calculating. Do we need all of NDCG@K / MAP@K / P@K  / recall@k ? (also I think we're doing this for multiple values of k ?)
> * What about just defaulting to cheaper metrics to calculate like AUC. With pair-wise losses this can be calculated basically for free (its the percentage of times the positive sample has a higher score than the negative sample_

@benfred you are using `model.compile(optimizer="adam", run_eagerly=True)` we run with `eagerly=False` to speed it up.

**Comment by benfred** [comment]
Thanks for the update @sararb ! those sound like good changes to get in

@rnyak - I will try out the KerasSequenceLoader.  Do we still have the perf regression with the change to merlin.io.Dataset though ? I thought the perf differences where in not using the Dataset object and just using a dask DDF.

> @benfred you are using model.compile(optimizer="adam", run_eagerly=True) we run with eagerly=False to speed it up.

Thats a good call out - per epoch time improvements almost 20x with that change (from 14 hours down to about 45 minutes for me). Its still pretty slow though. 

>     also doing this makes it faster:

I'm trying to get this going with the negative sampling, and I don't think that switching to a binary classification task will accomplish this though ? 

Is there a way to change the metrics with the default ItemRetrievalTask?  I see that the many of the tasks have a 'metrics' argument:  for instance the 'NextItemPredictionTask' has a metrics argument https://github.com/NVIDIA-Merlin/models/blob/5524f9f12eb7f3fedcd2659d166f023e8c5c72c0/merlin/models/tf/prediction/item_prediction.py#L594-L597 but I don't see the same for ItemRetrievalTask (and passing a metrics argument via kwargs also doesn't work for me).

**Comment by rnyak** [comment]
> 

@benfred  would that work for you ?

```
mf = mm.TwoTowerBlock(schema, query_tower= mm.MLPBlock([32, 64]))
model = mf.connect(mm.ItemRetrievalTask(samplers = [mm.InBatchSampler()], schema=schema, metrics=[]))
model.compile(optimizer="adam", run_eagerly=False)
train_loader = get_dataloader(nvt.Dataset(train_paths), shuffle=True)
losses = model.fit(train_loader, epochs=1)
```

**Comment by rnyak** [comment]
@benfred overall you are right. Training times are pretty long. I train a MF + Binary classification model in 14 min only for 1 epoch and only AUC metric.. this is too slow.

**Comment by gabrielspmoreira** [comment]
We have merged the #226 PR, which besides improving the performance of ranking metrics adds the `train_metrics_steps` argument to `model.fit()`, making it compute train metrics each N steps.
As a reference of performance, I am training a Two tower retrieval model with multiple user and item features and in-batch sampling (i.e. each positive item is paired with 256 sampled negative items (batch size). The snippet is below.

The **steps per second** with `train_metrics_steps=1` (default) is **31** and with `train_metrics_steps=100` is **35** (about **13% throughput improvement for this example)**

```python
mf = mm.TwoTowerBlock(schema, query_tower=mm.MLPBlock([[32]))
model = mf.connect(
    mm.ItemRetrievalTask(
        samplers=[mm.InBatchSampler()],
        schema=schema,
        metrics=[RecallAt(10), MRRAt(10), NDCGAt(10), AvgPrecisionAt(10), PrecisionAt(10)],
        loss="categorical_crossentropy",
        normalize=True,
    )
)
model.compile(optimizer="adam", run_eagerly=False)
train_loader = get_dataloader(nvt.Dataset(train_paths), batch_size=256, shuffle=False)
losses = model.fit(
    train_loader,
    epochs=10,
    train_metrics_steps=100,
)
````

**Comment by benfred** [comment]
I've tried this out with the latest improvements - and the training time for the original snippet I gave w/ movielens-25m has dropped from 14 hours an epoch down to about 3 minutes an epoch (about 280x faster).  I still feel like when can further work on performance improvements here - but I think we can close this issue
